### DISCUSSION TITLE
Calling "useUserStore(store)" in a Quasar boot file breaks the Pinia Persistedstate Plugin

### DISCUSSION THREAD

**Discussion by BenJackGill** [quasarframework/quasar#12736] [General - Components / Directives / etc] [answered]
I'm trying to use the [Pinia Persisted State Plugin](https://github.com/prazdevs/pinia-plugin-persistedstate) with Pinia in my Quasar app.

Out of the box everything works fine.

But when using a Quasar boot file the persisted state stops working. Refreshing the page wipes all the new values away.

I don't know why the boot file breaks the persisted state plugin, but I have narrowed the culprit down to a single line...

This is how I am adding the plugin to Pinia:

**`src/store/index.ts`**
```TypeScript
/* eslint-disable @typescript-eslint/no-unused-vars */
import { store } from 'quasar/wrappers';
import { createPinia, Pinia } from 'pinia';
import piniaPluginPersistedstate from 'pinia-plugin-persistedstate';

declare module '@quasar/app' {
  interface BootFileParams<TState> {
    store: Pinia;
  }
  interface PreFetchOptions<TState> {
    store: Pinia;
  }
}

declare module '@vue/runtime-core' {
  interface ComponentCustomProperties {
    $store: import('pinia').Pinia;
  }
}

export default store(function (_) {
  const pinia = createPinia();
  pinia.use(piniaPluginPersistedstate); // Pinia Plugin added here
  return pinia;
});
```

And this is what my Pinia store looks like:

**`src/store/user.ts`**
```TypeScript
import { defineStore } from 'pinia';

export const useUserStore = defineStore('user', {
  state: () => {
    return {
      user: {
        firstName: 'Mary',
      },
    };
  },
  persist: true, // Note that we are using a persisted state here
  actions: {
    setFirstName(firstName: string) {
      this.user.firstName = firstName;
      console.log('Name set to Pinia store: ', this.user.firstName);
    },
    getFirstName() {
      if (!this.user.firstName) {
        console.log('No name found in store. Setting "John" to Pinia store.');
        this.user.firstName = 'John';
        return this.user.firstName;
      } else {
        console.log('Name fetched from Pinia store: ', this.user.firstName);
        return this.user.firstName;
      }
    },
  },
});
```
Here is an example front-end page for fetching and setting the firstName:

**`src/pages/index.vue`**
```TypeScript
<template>
  <div>{{ firstName }}</div>
  <q-form @submit="handleFirstNameSubmit">
    <p>Change First Name</p>
    <q-input v-model="firstNameInput" filled outline />
    <q-btn label="Submit Name to Pinia Store" type="submit" />
  </q-form>
  <q-btn @click="handleFirstNameFetch" label="Fetch Name from Pinia Store" />
</template>

<script setup lang="ts">
import { ref } from 'vue';
import { useUserStore } from 'src/store/user';
const userStore = useUserStore();
const firstName = ref<string>();
const firstNameInput = ref<string>();
const handleFirstNameSubmit = () => {
  if (firstNameInput.value) {
    userStore.setFirstName(firstNameInput.value);
  }
};
const handleFirstNameFetch = () => {
  firstName.value = userStore.getFirstName();
};
</script>
```

Up to this point everything works fine.

I can set firstName to the Pinia store, refresh the page, and the new name is still in Pinia.

But when using `const userStore = useUserStore(store)` inside a boot file like the example below, the persisted state stops working:

**`src/boot/auth.ts`**
```TypeScript
import { boot } from 'quasar/wrappers';
import { useUserStore } from 'src/store/user';

export default boot(({ store }) => {
  const userStore = useUserStore(store);
  // Do some other authentication stuff, setting initial user store values etc, below here...
});
```

Any idea what's going on? And how to fix it?

I think this plugin is much cleaner than using LocalStorage so I would love to get it working with Quasar.

**Comment by yusufkandemir** [comment, marked as answer]
As you already noticed, it works without any problems in Vue files, but Pinia has a small limitation that affects outside Vue files(_e.g. boot files_):

https://pinia.vuejs.org/core-concepts/plugins.html#introduction
> Plugins are only applied to stores **created after `pinia` is passed to the app**, otherwise, they won't be applied.

**Pinia is now passed directly to the app before executing the boot files. So, this limitation is by-passed now. Available on `@quasar/app-vite-v1.0.0-beta.5`+ and `@quasar/app-webpack-v3.4.3`+**

~~The Pinia instance gets passed to the `app` after all the boot files are finished, so you need to be careful about that. Here are two examples, the first one doesn't work, the other one works:~~

<details>
  <summary>Outdated code samples to explain the old limitation</summary>

```js
export default boot(({ store, router }) => {
  // ❌ plugins will not be applied, because pinia is passed to app after all the boot files are finished
  const userStore = useUserStore(store);
  router.beforeEach((to, from) => {
    // use the store
  });
});
```

```js
export default boot(({ store, router }) => {
  router.beforeEach((to, from) => {
    // ✅ plugins will be applied, because router starts working after everything is initialized
    const userStore = useUserStore(store);
  });
});
```
</details>

**Reply by BenJackGill** [reply]
Very helpful once again! I now understand this much better.

Ultimatley I took the approach given by @metalsadman because it suited my needs more.

But I'm honestly shocked at the amount of passion and help you continue to provide to the Quasar community while maintaining this project.

I'm dead broke right now, but once I come into some coin I'm pouring it into Quasar donations!

A huge thanks for doing what you do :)

**Reply by yusufkandemir** [reply]
@BenJackGill I updated the answer to reflect the latest situation. Pinia is now passed directly to the app before executing the boot files. So, this limitation is by-passed now. Available on `@quasar/app-vite-v1.0.0-beta.5`+ and `@quasar/app-webpack-v3.4.3`+

**Reply by folamy** [reply]
> As you already noticed, it works without any problems in Vue files, but Pinia has a small limitation that affects outside Vue files(_e.g. boot files_):
> 
> https://pinia.vuejs.org/core-concepts/plugins.html#introduction
> 
> > Plugins are only applied to stores **created after `pinia` is passed to the app**, otherwise, they won't be applied.
> 
> **Pinia is now passed directly to the app before executing the boot files. So, this limitation is by-passed now. Available on `@quasar/app-vite-v1.0.0-beta.5`+ and `@quasar/app-webpack-v3.4.3`+**
> 
> ~The Pinia instance gets passed to the `app` after all the boot files are finished, so you need to be careful about that. Here are two examples, the first one doesn't work, the other one works:~
> 
> Outdated code samples to explain the old limitation
> ```js
> export default boot(({ store, router }) => {
>   // ❌ plugins will not be applied, because pinia is passed to app after all the boot files are finished
>   const userStore = useUserStore(store);
>   router.beforeEach((to, from) => {
>     // use the store
>   });
> });
> ```
> 
> ```js
> export default boot(({ store, router }) => {
>   router.beforeEach((to, from) => {
>     // ✅ plugins will be applied, because router starts working after everything is initialized
>     const userStore = useUserStore(store);
>   });
> });
> ```

@yusufkandemir, is there a way to initiate pinia store/state in prefetch()?

**Reply by rstoenescu** [reply]
@folamy Like already mentioned, Pinia store is available immediately after you're defining it in the default exporting function of /src/stores/index.js. So you can use your stores in preFetch too.

**Reply by folamy** [reply]
Thanks, I have already tried it, and it works!

**Comment by metalsadman** [comment]
Just started migrating some of my vuex modules to pinia and this plugin works fine for me, the only difference (apart from ts) is I created the pinia store in a boot file ie.
```js
// boot/pinia.js
import { boot } from 'quasar/wrappers'
import { createPinia } from 'pinia'
import piniaPluginPersistedstate from 'pinia-plugin-persistedstate'

export default boot(({ app }) => {
  const pinia = createPinia()
  pinia.use(piniaPluginPersistedstate)
  app.use(pinia)
})
```
and registered the boot in my quasar conf.js at a higher order in the boot array.

I just left the store/index.js as is, since I still need vuex to work for most stuff until i'm done migrating it all (it's quite huge so :/).

**Reply by BenJackGill** [reply]
Thanks! I took this approach because it seems to give the most flexibility and keeps things "modular".

In doing so I deleted my `src/store/index.ts` file and I'm wondering if I should also delete the `src/store/store-flag.d.ts` file that was auto generated when I first added `index.ts`?

I know that you are currently keeping your `index.ts`, but I thought you might have some insight regardless :)

**Reply by yusufkandemir** [reply]
@BenJackGill if you remove your `src/store/index.[js|ts]`, you lose the `store` parameter in boot files, `src/router/index.[js|ts]` file, etc. Having `src/store/store-flag.d.ts` gives the "illusion" of still having that parameter because of the types, so you should also remove that. Because you removed the `src/store` handling, you no longer can access Pinia stores in boot files, because you will no longer have access to `store`, thus can't call `useWhateverStore(store)`. One of the main reasons why `src/store` and `src/router` exists is being able to easily reference them in boot files, `preFetch`, etc. Another one would be handling SSR hydration and such under the hood, etc.

The idea behind this answer is to keep both Vuex and Pinia active while using the Vuex as the main store(_`store` param in boot files, etc._), to be able to progressively migrate from Vuex to Pinia. After fully migrating to Pinia, one would remove the boot file about Pinia, and move that logic to `src/store/index.[js|ts]`  to be able to get all the advantages. I am planning to mention both this migration path and Pinia's limitations about plugins in our Pinia docs.

**Reply by BenJackGill** [reply]
Ok thanks, I have now moved my Pinia logic from the boot file back into `src/store/index.ts`.

**Reply by tggm** [reply]
@metalsadman Did you change anything else? I'm trying to replicate this but the storage is lost on refresh. I noticed that it correctly set local storage values, but on refresh these are not being used to re-hydrate Pinia.

Also, you mentioned that you left store/index.js  as is. Does this mean that you're initializing pinia (createPinia() and pinia.use(piniaPluginPersistedstate) both on store/index.js and boot/pinia.js?

Thanks!

**Reply by tggm** [reply]
Nevermind. Store is only being populated when used. solved!
